### PR TITLE
Architecture overview

### PR DESCRIPTION
**Pull request by AlirezaSohofi**
# Description

Add a `Architecture overview` page to the doc with a complete data loading pipeline and it's equivalent UML diagram. You can use [mermaid live](https://mermaid.live/) to see the diagram.

Fixes # issue

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update
- [ ] Refactoring including code style reformatting 
- [ ] Other (please describe):

# Checklist:

- [ ] I have read the [contributing guideline doc](https://squirrel-core.readthedocs.io/en/latest/usage/code_of_conduct.html) (external contributors only)
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have kept the PR small so that it can be easily reviewed  
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] All dependency changes have been reflected in the pip requirement files.

### CODE REVIEW

File: docs/usage/architecture_overview.rst [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+            serialize(obj)
+            deserialize(obj)
+            serialize_shard_to_file(obj, fp)
+            deserialize_shard_from_file(fp)
+        }
+
+
+The relationships between these components and the methods they provide depends on the particular implementation of
+the abstract classes (i.e. :py:class:`Drive`, :py:class:`AbstractStore`, :py:class:`SquirrelSerializer`).
+For instance, an implementation of the :py:class:`Drive` may not need or choose to not use :py:class:`SquirrelStore`
```
Comment (pzdkn) [inline docs/usage/architecture_overview.rst:105]: Spelling error in Driver: `the abstract classes (i.e. :py:class:`Driver`, :py:class:`AbstractStore`, :py:class:`SquirrelSerializer`).`

File: docs/usage/architecture_overview.rst [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+* :ref:`driver`: performant and convenient read.
+* :ref:`iterstream`: :py:class:`Composable` is the foundational building block in the :ref:`iterstream` module which provides a mechanism to chain iterables, and a fluent api that includes methods such as `map`, `filter`, and `async_map`.
+* :ref:`store`: a key/value abstraction for reading and writing data to and arbitrary storage backend such as filesystem, object store, database, etc.
+
+These modules are designed in a way that `can` be
+used together, but it is not enforced in order to maximize flexibility. This may make it difficult to realize the
+intended and recommended way of combining squirrel primitives. Although there are many such ways already provided (and many more
+that can be implemented for specific use-cases), here we focus on one concrete example that captures the most common
+and most widely applicable pattern through a code snippet and its equivalent UML diagram.
+here is a complete data loading pipeline:
```
Comment (pzdkn) [inline docs/usage/architecture_overview.rst:16]: -> Here is a complete data loading pipeline:

File: docs/usage/architecture_overview.rst [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        .get_iter()  # Composable
+        .map(lambda x: transform(x)) # Composable
+        .filter(lambda x: filter_func(x)) # Composable
+    )
+    train_loader = DataLoader(train_data, batch_size=None)
+
+
+A :py:class:`Catalog` contains zero to many :py:class:`CatalogSource`, each of which may be retrieved by a an
+identifier (or a tuple of and identifier and the version, see :ref:`catalog` for more details). :py:class:`CatalogSource`
+contains all necessary information to instantiate an abject of type :py:class:`Driver`. :py:class:`Driver` may have
```
Comment (pzdkn) [inline docs/usage/architecture_overview.rst:36]: abject -> object

File: docs/usage/architecture_overview.rst [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+a method :py:meth:`get_iter` which returns an object of type :py:class:`Composable`
+(which belongs to :ref:`iterstream` module). `train_data` is an iterable that generates items lazily in a
+streaming way for minimal memory footprint.
+
+.. note::
+      To access available datasets using `Catalog.from_plugins()`, checkout `squirrel-dataset-core repository <https://github.com/merantix-momentum/squirrel-datasets-core>`_.
+
+The following diagram illustrating (a simplified and slightly idealized) view on the relationships between these
```
Comment (pzdkn) [inline docs/usage/architecture_overview.rst:44]: illustrating -> illustrates

**Comment by pzdkn** [comment]
Thanks Ali, it would looks quite nice and can maybe even act as an introduction?

File: docs/usage/architecture_overview.rst [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+    catalog = Catalog.from_plugins() # Catalog
+    train_data = (
+        catalog["imagenet"] # CatalogSource
+        .get_driver()  # Driver
+        .get_iter()  # Composable
+        .map(lambda x: transform(x)) # Composable
+        .filter(lambda x: filter_func(x)) # Composable
+    )
+    train_loader = DataLoader(train_data, batch_size=None)
```
Comment (AlpAribal) [inline docs/usage/architecture_overview.rst:31]: Additional explanation and/or pointers for some lines here would be useful (some points I can think of are: what does from_plugins do, what is a plugin, where is "imagenet" coming from, what kind of funcs are `transform` and `filter_func`). Maybe converting the following paragraph to a step-by-step explanation format would be better?

File: docs/usage/architecture_overview.rst (hunk 4 above)
Comment (AlpAribal) [inline docs/usage/architecture_overview.rst:44]:
```suggestion
The following diagram illustrates a (simplified and slightly idealized) view of the relationships between these
```

File: docs/usage/architecture_overview.rst (hunk 1 above)
Comment (AlpAribal) [inline docs/usage/architecture_overview.rst:105]:
```suggestion
For instance, an implementation of the :py:class:`Drive` may not need to or may choose not to use :py:class:`SquirrelStore`
```

**Review by AlpAribal — CHANGES_REQUESTED** [review]
Much needed summary, thank you, added few comments

**Comment by AlpAribal** [comment]
If you are using VSCode, I found the Esbonio (id: swyddfa.esbonio) extension quite useful for previewing sphinx docs.

File: docs/usage/architecture_overview.rst (hunk 5 above)
Comment (AlpAribal) [reply docs/usage/architecture_overview.rst:31]: I see, yes it is clear :+1:

File: docs/usage/architecture_overview.rst [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    from squirrel.catalog import Catalog
+
+    catalog = Catalog.from_plugins() # Catalog
+    train_data = (
+        catalog["imagenet"] # CatalogSource
+        .get_driver()  # Driver
+        .get_iter()  # Composable
+        .map(lambda x: transform(x)) # Composable
+        .filter(lambda x: filter_func(x)) # Composable
+    )
```
Comment (AlpAribal) [inline docs/usage/architecture_overview.rst:30]:
```suggestion
    train_data = (
        catalog["imagenet"]  # CatalogSource
        .get_driver()  # Driver
        .get_iter()  # Composable
        .map(lambda x: transform(x))  # Composable
        .filter(lambda x: filter_func(x))  # Composable
    )  # Composable
```

**Review by AlpAribal — APPROVED** [review]
LGTM

File: docs/usage/architecture_overview.rst [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+and most widely applicable pattern through a code snippet and its equivalent UML diagram.
+Here is a complete data loading pipeline:
+
+.. code-block:: python
+
+    from squirrel.catalog import Catalog
+
+    catalog = Catalog.from_plugins() # Catalog
+    train_data = (
+        catalog["imagenet"] # CatalogSource
```
Comment (AlpAribal) [inline docs/usage/architecture_overview.rst:24]:
```suggestion
    catalog = Catalog.from_plugins()  # Catalog
    train_data = (
        catalog["imagenet"]  # CatalogSource
```

File: docs/usage/architecture_overview.rst [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        class MessagepackDriver {
+            string name
+            SquirrelStore store
+
+            get(key) Iterable~Dict~
+            keys() List~string~
+            get_iter() Composable
+        }
+
+        %% realiazation
```
Comment (AlpAribal) [inline docs/usage/architecture_overview.rst:74]:
```suggestion
        %% realization
```

**Review by AlpAribal — APPROVED** [review]

**Pull request merged**